File: matlab/+py_ionization/+charge_exchange/production_rate.m
function rate_s_1 = production_rate(energy_eV, ion_def, varargin)
%PRODUCTION_RATE Compute a charge exchange production rate.
%
% rate_s_1 = py_ionization.charge_exchange.production_rate(energy_eV, ion_def, "ReactionType", "HpH>HpH")
%
% Name-value options
%   ReactionType  Compact reaction key, for example "HpH>HpH".
%   Ion           Incident ion species.
%   Neutral       Neutral target species.
%   Dphi          Azimuthal width in radians, default 2*pi.
%   ThetaRange    Theta limits in radians, default [-pi/4, pi/4].
%   ThetaType     "auto", "elevation", or "polar", default "auto".
%   MaxEnergyEV   Energies >= this value are ignored, default 4000 eV.

p = inputParser;
p.addParameter("ReactionType", "", @(x) ischar(x) || isstring(x));
p.addParameter("Ion", "", @(x) ischar(x) || isstring(x));
p.addParameter("Neutral", "", @(x) ischar(x) || isstring(x));
p.addParameter("Dphi", 2*pi, @(x) isnumeric(x) && isscalar(x));
p.addParameter("ThetaRange", [-pi/4, pi/4], @(x) isnumeric(x) && numel(x) == 2);
p.addParameter("ThetaType", "auto", @(x) ischar(x) || isstring(x));
p.addParameter("MaxEnergyEV", 4000.0, @(x) isnumeric(x) && (isempty(x) || isscalar(x)));
p.parse(varargin{:});

energy = double(energy_eV(:).');
if any(energy <= 0)
    error("py_ionization:charge_exchange:BadEnergy", ...
        "energy_eV must be positive.");
end

flux = double(ion_def);
scalar_input = isvector(flux);
if scalar_input
    flux = reshape(flux, 1, []);
elseif size(flux, 2) ~= numel(energy) && size(flux, 1) == numel(energy)
    flux = flux.';
end
if size(flux, 2) ~= numel(energy)
    error("py_ionization:charge_exchange:SizeMismatch", ...
        "ion_def must have the same number of energy bins as energy_eV.");
end

[ion, neutral] = infer_reaction_species(p.Results.ReactionType, p.Results.Ion, p.Results.Neutral);
if ion == "" || neutral == ""
    error("py_ionization:charge_exchange:MissingReaction", ...
        "Specify either ReactionType or both Ion and Neutral.");
end

dE = py_ionization.energy_bin_width(energy);
[domega, ~] = py_ionization.solid_angle_from_theta_range( ...
    p.Results.ThetaRange, "Dphi", p.Results.Dphi, "ThetaType", p.Results.ThetaType);

energy_mat = repmat(energy, size(flux, 1), 1);
differential_particle_flux = flux ./ energy_mat;
if ~isempty(p.Results.MaxEnergyEV)
    differential_particle_flux(energy_mat >= p.Results.MaxEnergyEV) = nan;
end

sigma = py_ionization.charge_exchange.cross_section(energy, "Ion", ion, "Neutral", neutral);
integrand = differential_particle_flux .* sigma .* domega .* dE;
integrand(~isfinite(integrand)) = 0;
rate_s_1 = sum(integrand, 2);

if scalar_input
    rate_s_1 = rate_s_1(1);
end
end

function [ion, neutral] = infer_reaction_species(reaction_type, ion, neutral)
ion = string(ion);
neutral = string(neutral);
key = lower(regexprep(string(reaction_type), "\s+", ""));

if key == ""
    return
end

switch key
    case "hpo>oph"
        inferred_ion = "H+"; inferred_neutral = "O";
    case "hph>hph"
        inferred_ion = "H+"; inferred_neutral = "H";
    case "oph>hpo"
        inferred_ion = "O+"; inferred_neutral = "H";
    case "opo>opo"
        inferred_ion = "O+"; inferred_neutral = "O";
    case "hpn2>n2ph"
        inferred_ion = "H+"; inferred_neutral = "N2";
    case "hpo2>o2ph"
        inferred_ion = "H+"; inferred_neutral = "O2";
    otherwise
        error("py_ionization:charge_exchange:UnsupportedReactionType", ...
            "Unsupported ReactionType: %s", reaction_type);
end

if ion == ""
    ion = inferred_ion;
end
if neutral == ""
    neutral = inferred_neutral;
end
end

File: py_ionization/charge_exchange/production_rate.py
from __future__ import annotations

import numpy as np

from py_ionization._utils import energy_bin_width, solid_angle_from_theta_range

from .cross_section import charge_exchange_cross_section


_REACTION_MAP = {
    "hpo>oph": ("H+", "O"),
    "hph>hph": ("H+", "H"),
    "oph>hpo": ("O+", "H"),
    "opo>opo": ("O+", "O"),
    "hpn2>n2ph": ("H+", "N2"),
    "hpo2>o2ph": ("H+", "O2"),
}


def _reaction_to_species(reaction_type: str) -> tuple[str, str]:
    key = str(reaction_type).lower().strip().replace(" ", "")
    try:
        return _REACTION_MAP[key]
    except KeyError as exc:
        supported = ", ".join(sorted(_REACTION_MAP))
        raise ValueError(f"Unsupported reaction_type {reaction_type!r}. Supported keys: {supported}") from exc


def charge_exchange_rate(
    energy_eV,
    ion_def,
    *,
    reaction_type: str | None = None,
    ion: str | None = None,
    neutral: str | None = None,
    dphi: float = 2 * np.pi,
    theta_range=(-np.pi / 4, np.pi / 4),
    theta_type: str = "auto",
    max_energy_eV: float | None = 4000.0,
):
    """Compute a charge exchange production rate from ion differential energy flux.

    Parameters
    ----------
    energy_eV
        1D ion energy bins in eV.
    ion_def
        Ion differential energy flux in eV cm^-2 s^-1 sr^-1 eV^-1.
    reaction_type
        Optional compact reaction key, for example ``"HpH>HpH"``. If this is
        supplied, ``ion`` and ``neutral`` are inferred unless explicitly set.
    ion, neutral
        Incident ion and neutral target species.
    dphi, theta_range, theta_type
        Angular integration geometry.
    max_energy_eV
        Energies greater than or equal to this value are ignored. The default
        is 4000 eV.

    Returns
    -------
    rate_s_1
        Charge exchange rate in s^-1 for one neutral target particle.

    Formula
    -------
    ``rate = integral (DEF / E) sigma(E) dE dOmega``.
    """
    energy = np.asarray(energy_eV, dtype=float)
    if energy.ndim != 1:
        raise ValueError("energy_eV must be a 1D array.")
    if np.any(energy <= 0):
        raise ValueError("energy_eV must be positive.")

    flux = np.asarray(ion_def, dtype=float)
    scalar_time = flux.ndim == 1
    if scalar_time:
        flux = flux[None, :]
    elif flux.ndim != 2:
        raise ValueError("ion_def must be 1D or 2D.")
    if flux.shape[1] != energy.size:
        raise ValueError("ion_def must have the same number of energy bins as energy_eV.")

    if reaction_type is not None:
        inferred_ion, inferred_neutral = _reaction_to_species(reaction_type)
        ion = inferred_ion if ion is None else ion
        neutral = inferred_neutral if neutral is None else neutral
    if ion is None or neutral is None:
        raise ValueError("Specify either reaction_type or both ion and neutral.")

    dE = energy_bin_width(energy)[None, :]
    energy_mat = energy[None, :]
    domega, _ = solid_angle_from_theta_range(theta_range, dphi=dphi, theta_type=theta_type)
    differential_particle_flux = flux / energy_mat
    if max_energy_eV is not None:
        differential_particle_flux = np.where(energy_mat >= max_energy_eV, np.nan, differential_particle_flux)

    sigma = np.asarray(charge_exchange_cross_section(energy, ion=ion, neutral=neutral), dtype=float)
    rate = np.nansum(differential_particle_flux * sigma[None, :] * domega * dE, axis=1)
    if scalar_time:
        return float(rate[0])
    return rate


cex_prod_rate = charge_exchange_rate
charge_exchange_prod_rate = charge_exchange_rate
production_rate = charge_exchange_rate

__all__ = ["charge_exchange_rate", "cex_prod_rate", "charge_exchange_prod_rate", "production_rate"]
